api: add follow param to file stream endpoint

tgross opened the pull request:
The `/v1/client/fs/stream endpoint` supports tailing a file by writing chunks out as they come in. But not all browsers [support streams](https://caniuse.com/#feat=streams) (ex IE11) so we need to be able to tail a file without streaming.

The fs stream and logs endpoint use the same implementation for filesystem streaming under the hood, but the fs stream always passes the `follow` parameter set to true. This adds the same toggle to the fs stream endpoint that we have for logs. It defaults to true for backwards compatibility.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=6049)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=6049)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/nomad?pullRequest=6049).</sub>

notnoop commented on command/agent/fs_endpoint.go:
```diff
@@ -194,11 +194,14 @@ func (s *HTTPServer) FileCatRequest(resp http.ResponseWriter, req *http.Request)
 //           applied. Defaults to "start".
 func (s *HTTPServer) Stream(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
 	var allocID, path string
+	var follow bool = true
```
This one can be and keep it closer to the block with `followStr`.  It's more conventional in go to keep variable initializations closer to where they are first used.  This function uses this pattern for `offset`, but doesn't for `allocID` and `path` 🤷‍♂

```suggestion
	follow := true
```

notnoop commented on command/agent/fs_endpoint_test.go:
```diff
@@ -342,6 +342,45 @@ func TestHTTP_FS_Cat(t *testing.T) {
 }
 
 func TestHTTP_FS_Stream(t *testing.T) {
```
Given that this tests the non-default case, I would call the follow=false in the function name?  So maybe 

```suggestion
func TestHTTP_FS_Stream_NoFollow(t *testing.T) {
```

notnoop commented on command/agent/fs_endpoint_test.go:
```diff
@@ -342,6 +342,45 @@ func TestHTTP_FS_Cat(t *testing.T) {
+		respW := testutil.NewResponseRecorder()
+		go func() {
+			_, err = s.Server.Stream(respW, req)
+			require.Nil(err)
```
I think we need to borrow the `close(doneCh)` here as well but we need to check later that it does close, so reverse the logic in other follow=true case: https://github.com/hashicorp/nomad/blob/293d2123c2ae448329954f52747c64a65a994516/command/agent/fs_endpoint_test.go#L504-L508 .

FWIW, this and the existing test aren't proper in using `require.` assertions in a goroutine.  They don't behave well, but don't worry about it now :).

tgross replied:
> FWIW, this and the existing test aren't proper in using require. assertions in a goroutine. They don't behave well, but don't worry about it now :).

Ok I'll leave this for now but yeah that sounds like it'd cause some interesting behavior on failure. 😁

tgross commented:
@notnoop @cgbaker I've addressed your comments, but I just wanted to have a second look before merging as this is my first PR (including the doc & changelog changes).

cgbaker commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 IMPROVEMENTS:
  * agent: allow the job GC interval to be configured [[GH-5978](https://github.com/hashicorp/nomad/issues/5978)]
+ * api: add follow parameter to file streaming endpoint to support older browsers [[GH-6049](https://github.com/hashicorp/nomad/pull/6049]
```
i think this is missing a closing parenthesis
these are easy to mess up by copy-pasting. we tend to use `[GH-####]` and then run `make changelogfmt`

tgross replied:
Ah, thanks for catching that. Done.

I'll merge this once CI gives the :+1: again.
